### PR TITLE
Implements -J and  -I

### PR DESCRIPTION
**Pull request by meow464**
I will rebase before merge. At least locally I can build the cross compilation branch with ninja.

Please check if any of the renames `CompilerOptions &compiler_options` -> `CompilerOptions& compiler_options` looks wrong. In principle it should all work because `&compiler_options` takes a reference and  `CompilerOptions&` expects a reference (or at least that is my understanding, C++ syntax can be confusing).

### CODE REVIEW

**Comment by meow464** [comment]
I will check why windows is failing some time this week.

**Comment by certik** [comment]
The Windows error seems to be:
```
D:\a\lfortran\lfortran\lfortran-0.16.0-640-ge77e7ff49\src\libasr\asr_utils.cpp(256): error C2665: 'LFortran::read_file': none of the 2 overloads could convert all the argument types
D:\a\lfortran\lfortran\lfortran-0.16.0-640-ge77e7ff49\src\libasr\..\libasr/utils.h(53): note: could be 'bool LFortran::read_file(const std::string &,std::string &)'
D:\a\lfortran\lfortran\lfortran-0.16.0-640-ge77e7ff49\src\libasr\asr_utils.cpp(256): note: while trying to match the argument list '(std::filesystem::path, std::string)'
```

File: src/libasr/asr_utils.cpp [hunk 1]
```diff
@@ -236,19 +238,30 @@ void set_intrinsic(ASR::TranslationUnit_t* trans_unit) {
...
-    }
-    return asr;
+    CompilerOptions& compiler_options = get_compiler_options();
+    std::filesystem::path filename {msym + ".mod"};
+    std::vector<std::filesystem::path> mod_files_dirs;
+
+    // inserting rl_path first is important to keep the previous behavior
+    // whether rl_path should be used to search for mod files is another matter
+    mod_files_dirs.push_back( std::filesystem::path{rl_path} );
+    mod_files_dirs.push_back( compiler_options.mod_files_dir );
```
Comment (certik) [inline src/libasr/asr_utils.cpp:248]: I would pass `mod_files_dir` as an argument to `find_and_load_module`.

File: src/libasr/utils2.cpp [hunk 2]
```diff
@@ -11,6 +11,12 @@
 
 namespace LFortran {
 
+static CompilerOptions compiler_options;
```
Comment (certik) [inline src/libasr/utils2.cpp:14]: This is effectively a global variable. I would recommend not to use global variables. Just pass it as a const argument.

File: src/bin/lfortran.cpp [hunk 3]
```diff
@@ -363,7 +364,7 @@ int prompt(bool verbose)
 }
 #endif
 
-int emit_prescan(const std::string &infile, CompilerOptions &compiler_options)
+int emit_prescan(const std::string &infile, CompilerOptions& compiler_options)
```
Comment (certik) [inline src/bin/lfortran.cpp:367]: Send these formatting changes as a separate PR. That way it's easier to review.

**Comment by certik** [comment]
Btw, I think `CompilerOptions &compiler_options` is equivalent to `CompilerOptions& compiler_options`, there is no difference.

**Comment by certik** [comment]
The actual change I think is fine. 

In general, I recommend to just send a PR with the fix. Everything else, like refactoring, syntax formatting, etc. should be done in separate PRs. That way it is easier to review and get in.

File: src/libasr/asr_utils.cpp (hunk 1 above)
Comment (meow464) [reply src/libasr/asr_utils.cpp:248]: The problem is that `find_and_load_module` is many, many calls deep. It gets called by `load_module` in 4 places, `load_module` in turn gets called by a bunch of other functions and IIRC there are 4-5 levels of nesting, resulting in something like ~15 places where `compiler_options` would need to be passed and not used. Some of those places are in AST code that in principle should not need to know about compiler options.

I'm not a fan of global state either and I prefer my functions to be as pure as possible but this amount of passing parameters around also adds a lot of complexity (and `compiler_options` is passed as a mutable reference anyways). 

What I think would be an interesting solution is: `CompilerOptions` becomes a singleton class and the options are write-once protected by a setter method. We get simpler code and keep the purity of `CompilerOptions` users.

File: src/libasr/asr_utils.cpp [hunk 4]
```diff
@@ -236,19 +238,30 @@ void set_intrinsic(ASR::TranslationUnit_t* trans_unit) {
...
+    for (std::filesystem::path p : mod_files_dirs) {
+      std::string modfile;
+
+      if (read_file(p / filename, modfile)) {
+	ASR::TranslationUnit_t *asr = load_modfile(al, modfile, false, symtab);
+	if (intrinsic) {
+	  set_intrinsic(asr);
+	}
+	return asr;
+      }
```
[the remark this answers is not in the document]
Comment (meow464) [reply src/libasr/asr_utils.cpp:263]: I swear I take a lot of care with formatting, emacs is playing tricks on me.

**Comment by meow464** [comment]
I think having `--verbose` in the CI call to `cmake --build` would be helpful, I was going to install windows server in a VM today or tomorrow.

File: src/libasr/asr_utils.cpp (hunk 1 above)
Comment (meow464) [reply src/libasr/asr_utils.cpp:248]: `rl_path` could be in this immutable global too.

Comment (meow464) [reply src/libasr/asr_utils.cpp:248]: Initially I had considered a singleton class with stters and getters but this would require a lot of refactoring.

Comment (meow464) [reply src/libasr/asr_utils.cpp:248]: I made an attempt at passing CompilerOptions as an argument and there already are 30 functions handling it for no reason and many more remaining, that's a lot of added complexity. https://github.com/meow464/lfortran/tree/implement-include-and-mod-dir-by-argument

**Comment by certik** [comment]
I'll try to implement it using the approach I suggested soon, I think there is a way to do it.

Besides that, everything else works for you?

**Comment by meow464** [comment]
Besides that, yes. At least locally the cross compilation pull request builds.

**Comment by certik** [comment]
Awesome, thanks. I marked it as my TODO list. I will get to it soon.

**Comment by meow464** [comment]
I'm back working on this. Should be finished next week. Will probably do -L too.

**Comment by certik** [comment]
Awesome, thanks. I am here to help or answer questions.

**Comment by certik** [comment]
Fixed by #930.
